streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -541,10 +555,52 @@ public void suspend() {
      *                               or if the task producer got fenced (EOS)
      */
     // visible for testing
-    void suspend(final boolean clean) {
+    void suspend(final boolean clean,
+                 final boolean isZombie) {
         closeTopology(); // should we call this only on clean suspend?
         if (clean) {
-            commit(false);
+            TaskMigratedException taskMigratedException = null;
+            try {
+                commit(false);
+            } finally {
+                if (eosEnabled) {
+                    try {
+                        recordCollector.close();
+                    } catch (final ProducerFencedException e) {
+                        taskMigratedException = new TaskMigratedException(this, e);
+                    } finally {
+                        producer = null;
+                    }
+                }
+            }
+            if (taskMigratedException != null) {
+                throw taskMigratedException;
+            }
+        } else {
+            if (eosEnabled) {
+                try {
+                    if (!isZombie && transactionInFlight) {
+                        producer.abortTransaction();
+                    }
+                    transactionInFlight = false;
+                } catch (final ProducerFencedException ignore) {
+                    /* TODO
+                     * this should actually never happen atm as we guard the call to #abortTransaction
+                     * -> the reason for the guard is a "bug" in the Producer -- it throws IllegalStateException
+                     * instead of ProducerFencedException atm. We can remove the isZombie flag after KAFKA-5604 got
+                     * fixed and fall-back to this catch-and-swallow code
+                     */
+
+                    // can be ignored: transaction got already aborted by brokers/transactional-coordinator if this happens
+                }
+            }
+            try {
+                if (!isZombie) {
+                    recordCollector.close();
```

Comment:
Should we avoid calling `recordCollector.close()` unless `eosEnabled`?

It seems like this block used to be guarded by `if(eosEnabled)`, but it's not anymore.

Reply:
Good catch!